parse/grammar/KuneiformParser.g4

```diff
@@ -0,0 +1,378 @@
+;
+
+// action statements can only be 3 things:
+// 1. a sql statement
+// 2. a local action/procedure call.
+// 3. an extension call
+action_statement:
+    sql_statement # sql_action
+    | IDENTIFIER LPAREN (procedure_expr_list)? RPAREN # local_action
+    | (variable_list EQUALS)? IDENTIFIER PERIOD IDENTIFIER LPAREN (procedure_expr_list)? RPAREN # extension_action
+;
+
+/*
+    This section includes parser rules for procedures
+*/
+
+// procedure_block is the top-level rule for a procedure.
+procedure_block:
+    statement*
+;
+
+procedure_expr:
+    literal type_cast? # literal_procedure_expr
+    | procedure_function_call type_cast? # function_call_procedure_expr
+    | variable type_cast? # variable_procedure_expr
+    | LBRACKET (procedure_expr_list)? RBRACKET type_cast? # make_array_procedure_expr
+    | procedure_expr LBRACKET procedure_expr RBRACKET type_cast? # array_access_procedure_expr
+    | LPAREN procedure_expr RPAREN type_cast? # paren_procedure_expr
+    | procedure_expr PERIOD IDENTIFIER type_cast? # field_access_procedure_expr
+    | procedure_expr (EQUALS | EQUATE | NEQ | LT | LTE | GT | GTE) procedure_expr # comparison_procedure_expr
+    | (MINUS|PLUS|EXCL) procedure_expr # unary_procedure_expr
+    // setting precedence for arithmetic operations:
+    | procedure_expr CONCAT procedure_expr # procedure_expr_arithmetic
+    | procedure_expr (STAR | DIV | MOD) procedure_expr # procedure_expr_arithmetic
+    | procedure_expr (PLUS | MINUS) procedure_expr # procedure_expr_arithmetic
+;
+
+procedure_expr_list:
+    procedure_expr (COMMA procedure_expr)*
+;
+
+statement:
+    VARIABLE type SCOL # stmt_variable_declaration
+    // stmt_procedure_call must go above stmt_variable_assignment due to lexer ambiguity
+    | ((variable_or_underscore) (COMMA (variable_or_underscore))* ASSIGN)? procedure_function_call SCOL # stmt_procedure_call
+    | VARIABLE ASSIGN procedure_expr SCOL # stmt_variable_assignment
+    | VARIABLE type? ASSIGN procedure_expr SCOL # stmt_variable_assignment_with_declaration
+    | FOR receiver=VARIABLE IN (range|target_variable=variable|sql_statement) LBRACE statement* RBRACE # stmt_for_loop
+    | IF if_then_block (ELSEIF if_then_block)* (ELSE LBRACE statement* RBRACE)? # stmt_if
+    | sql_statement SCOL # stmt_sql
+    | BREAK SCOL # stmt_break
+    | RETURN (procedure_expr_list|sql_statement) SCOL # stmt_return
+    | RETURN NEXT procedure_expr_list SCOL # stmt_return_next
+;
+
+// unfortunately necessary to preserve order in generated code
+variable_or_underscore:
+    VARIABLE | UNDERSCORE
+;
+
+procedure_function_call:
+    IDENTIFIER LPAREN (procedure_expr_list)? RPAREN #normal_call_procedure
+    | IDENTIFIER LBRACKET dbid=procedure_expr COMMA procedure=procedure_expr RBRACKET LPAREN (procedure_expr_list)? RPAREN #foreign_call_procedure
+;
+
+if_then_block:
+    procedure_expr LBRACE statement* RBRACE
+;
+
+// range used for for loops
+range:
+    procedure_expr RANGE procedure_expr
```

Comment:
Well, my point is we might implicitly allow some syntax, which are not allowed by postgres/plpgsql, but are allow by parser/analyzer

Reply:
Yeah. I do my best to guard for these, but it is quite hard. I'm unsure if it is possible to guard against all of them because of how weird plpgsql syntax is. We might need to make this a postgres extension.

Reply:
I'm not sure how a postgres extension can help, but better than not.
We'll see as we write more test cases, I've tried to write any test yet. I'll try after you have a test setup